src/box/lua/schema.lua

```diff
@@ -632,6 +648,104 @@ box.schema.space.alter = function(space_id, options)
     _space:replace(tuple)
 end
 
+local function check_upgrade_mode(mode)
+    if mode ~= 'test' and mode ~= 'notest' and mode ~= 'upgrade' then
+        box.error(box.error.ILLEGAL_PARAMS,
+            "upgrade mode should be 'test|notest|upgrade' but got "..mode)
+    end
+end
+
+local function space_upgrade_run(space_id, mode, func_id, format)
+    local _space_upgrade = box.space._space_upgrade
+    local su = _space_upgrade:get({space_id})
+
+    local rc
+    if mode ~= "notest" then
+        -- It's OK if "test" entry will be replicated, it doesn't affect
+        -- anything: just easy way to parse format and fill in metadata.
+        --
+        _space_upgrade:replace({space_id, "test", func_id, format})
+        log.info("Starting :upgrade() test...")
+        rc = builtin.space_upgrade_test(space_id)
+        if rc ~= 0 then box.error() end
+        -- Test should provide cleanup of _space_upgrade entry.
+        assert(box.space._space_upgrade:get({space_id}) == nil)
+        log.info(":upgrade() test is finished")
+    end
+
+    if mode == "test" then return end
+
+    -- Firstly we should insert tuple to _space_upgrade to update
+    -- upgrade status. Status allows us to manage alter locks during
+    -- space lock and skip tuple verification.
+    --
+    _space_upgrade:replace({space_id, "inprogress", func_id, format})
+    log.info("Inserted entry to _space_upgrade")
+    -- After successful update of _space_upgrade we can alter space
+    -- (format and/or opts) via space:alter(). Even if format is not changed,
+    -- we should process insertion to _space in order to set alter lock on
+    -- space.
+    --
+    local options = {}
+    options.format = format
+    local _, err = pcall(box.schema.space.alter, space_id, options)
+    if err ~= nil and not su then
+        -- At this moment entry in _space_upgrade can be already committed,
+        -- so to make this function behave "consistently" let's remove
+        -- corresponding tuple.
+        --
+        box.space._space_upgrade:delete(space_id)
+        box.error(err)
+    end
+    log.info("Space format has been changed")
+
+    rc = builtin.space_upgrade(space_id)
```

Comment:
Yes, we need. In my first implementation I tried to place upgrade meta right in _space, but code & overall logic turned out to be terrible for many reasons. Firstly, to update status (which is frequent operation) we have to proceed all dll routine. Trigger body is a bit overloaded even now, but with these updates it'll become awful (for example, in case mode is "test" - we shouldn't update format). Secondly _space format itself will become overloaded as well - for upgrade we store status, uuid, function id.